db/db.go

```diff
@@ -168,154 +169,193 @@ func newDB(n app.Net, id thread.ID, options *NewDBOptions) (*DB, error) {
 		log.Fatalf("unable to connect app: %s", err)
 	}
 	d.connector = connector
-
 	return d, nil
 }
 
+// managedDatastore returns whether or not the datastore is
+// being wrapped by an external datastore.
+func managedDatastore(ds ds.Datastore) bool {
+	_, ok := ds.(kt.KeyTransform)
+	return ok
+}
+
 // reCreateCollections loads and registers schemas from the datastore.
 func (d *DB) reCreateCollections() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
 	results, err := d.datastore.Query(query.Query{
 		Prefix: dsDBSchemas.String(),
 	})
 	if err != nil {
 		return err
 	}
 	defer results.Close()
-
 	for res := range results.Next() {
 		name := ds.RawKey(res.Key).Name()
-
 		schema := &jsonschema.Schema{}
 		if err := json.Unmarshal(res.Value, schema); err != nil {
 			return err
 		}
-
-		var indexes map[string]IndexConfig
+		c, err := newCollection(name, schema, d)
+		if err != nil {
+			return err
+		}
+		var indexes map[string]Index
 		index, err := d.datastore.Get(dsDBIndexes.ChildString(name))
 		if err == nil && index != nil {
-			_ = json.Unmarshal(index, &indexes)
-		}
-
-		indexValues := make([]IndexConfig, len(indexes))
-		for _, value := range indexes {
-			indexValues = append(indexValues, value)
```

Comment:
This was a bug that resulted in the values being appended _to the end_ of the slice which already has length > 0. Looks like it was resulting in some blank indexes being persisted... not too harmful, but something to keep an eye on when we reload an existing repo with the new code.

Reply:
Good catch!

Reply:
Okay, playing on the safe side... I added a check here during re-hydration: https://github.com/textileio/go-threads/pull/362/commits/014b3363ade213f843cb54bbe70a3421a35fd26b